### PR TITLE
Extend lifetime of GPUExternalTexture imported from video element

### PR DESCRIPTION
**Pull request by kainino0x**
Original PR: #1666
Discussion: #2124

The change proposed here destroys "stale" textures (those which no longer represent the video's current frame) just **before** rAF callbacks (and before rVFC callbacks, which occur before rAF callbacks, if applicable, because once rVFC fires there's no need for the old texture). It exposes a readonly attribute "expired" which the application can check to know if it should re-import.

Notably, expiry occurs even if the texture was imported _just_ prior to rVFC/rAF. This means it's possible to import on an event (say, "ontimeupdate") and have the texture become invalid before getting a chance to use it. I think this is OK because (1) applications should be importing later, inside their rendering task (rAF or rVFC callback), and (2) the strict timing requirements should mean developers find any such errors immediately.
Alternatively there could be a provision that keeps such textures alive a bit longer, but this could result in applications accidentally introducing an extra (video) frame of latency unnecessarily.

***
<a href="https://pr-preview.s3.amazonaws.com/kainino0x/gpuweb/pull/2302.html" title="Last updated on Nov 30, 2021, 12:19 AM UTC (52baf82)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/2302/baa0d23...kainino0x:52baf82.html" title="Last updated on Nov 30, 2021, 12:19 AM UTC (52baf82)">Diff</a>

***
<a href="https://pr-preview.s3.amazonaws.com/kainino0x/gpuweb/pull/2302.html" title="Last updated on Feb 18, 2022, 12:10 AM UTC (a2d998f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/2302/0005996...kainino0x:a2d998f.html" title="Last updated on Feb 18, 2022, 12:10 AM UTC (a2d998f)">Diff</a>

### CODE REVIEW

**Review by Kangz — APPROVED** [review]
LGTM but don't we need something on the `GPUExternalTexture` to know if it is still alive?

**Comment by kainino0x** [comment]
You're right, I was cranking out spec PRs and totally forgot about that! Will mark as draft until that's added.

**Comment by toji** [comment]
LGTM as well! (Minor added bonus to this approach I haven't seen anybody comment on: It could reduce BindGroup churn.)

As for notifying developers that the texture is still alive, seems like a promise would be appropriate?

```js
function updateVideoTexture() {
  videoTexture = gpuDevice.importExternalTexture({ source: video });
  videoTexture.expired.then(updateVideoTexture);
}
```

**Comment by kainino0x** [comment]
> As for notifying developers that the texture is still alive, seems like a promise would be appropriate?

This would be very similar to requestVideoFrameCallback (rVFC) but phrased as a promise instead. However since it's tied to the end of the frame lifetime instead of the beginning, it could lead to developers to re-importing the video when the previous frame expires rather than when they need it, which could result in <1 animation frame of latency when the video advances between the beginning and end of the animation frame, because video frame advancement is totally desynchronized with the animation loop AFAIU. E.g. with both video and animation updating at the same rate, say 60Hz:

externalTexture.expired.then:
- Video element advances to frame 2
- rAF callback, app renders frame 1
- Expire imported frame 1
- Page renders
- expired.then fires, app imports frame 2
- Video element advances to frame 3
- rAF callback <- using frame 2 instead of frame 3

rVFC:
- Video element advances to frame 2, rVFC fires, app imports frame 2
- rAF callback renders frame 2
- Expire imported frame 1
- Page renders
- Video element advances to frame 3, rVFC fires, app imports frame 3
- rAF callback renders frame 3

That said, working through this exercise made me realize that a synchronous isExpired that changes after rAF callbacks is even worse:

externalTexture.isExpired:
- Video element advances to frame 2
- rAF callback, app checks if frame 1 is still valid; it is, so it renders frame 1
- Expire imported frame 1
- Page renders
- Video element advances to frame 3
- rAF callback, app checks if frame 1 is still valid; it isn't, so imports and renders frame 3
- Expire imported frame 2
- Page renders
- Video element advances to frame 4
- rAF callback, app checks if frame 3 is still valid; it is, so it renders frame 3

In conclusion, I need to think about this some more.

**Comment by kainino0x** [comment]
TIL requestVideoFrameCallback (as drafted) actually occurs during "update the rendering", not decoupled from page animation frames/rAF.

https://wicg.github.io/video-rvfc/#video-rvfc-procedures

**Comment by kainino0x** [comment]
Imitating the lifetimes in rVFC would fix it:

externalTexture.isExpired where it changes BEFORE rAF (before or after rVFC?):
- Video element advances to frame 2
- Expire imported frame 1
- rAF callback, app checks if frame 1 is still valid; it isn't, so imports and renders frame 2 <- up-to-date
- Page renders
- Video element advances to frame 3
- Expire imported frame 2
- rAF callback, app checks if frame 2 is still valid; it isn't, so imports and renders frame 3 <- up-to-date
- Page renders

Though we might still want to guarantee that an imported frame lasts until the next end-of-rAF in case it was imported in some other async code. (This doesn't affect this case study.)

**Comment by kainino0x** [comment]
Updated description with latest changes.

**Comment by Kangz** [comment]
Just to make sure I completely understand https://github.com/gpuweb/gpuweb/pull/2302#issuecomment-976100234, see is what would happen? (it sounds good)

 - Video element advances to frame 2
 - Update the rendering
   - Expire imported frame 1
   - rAF callback, app checks if frame 1 is still valid; it isn't, so imports and renders frame 2 <- up-to-date render
   - Page renders
 - Video element advances to frame 3
 - Update the rendering
   - Expire imported frame 2
   - rAF callback, app checks if frame 2 is still valid; it isn't, so imports and renders frame 3 <- up-to-date render
   - Page renders

**Comment by kainino0x** [comment]
Yup that looks correct.

Technically the video element can advance at any time (I think). I tried to diagram this out yesterday but the resulting diagram was kind of incomprehensible. The rules as written in the spec make more sense. (Just before rVFC/rAF, textures expire if they're now outdated.)

**Comment by kainino0x** [comment]
Editors to decide exactly what input we need from the group, or specific individuals.

**Comment by kainino0x** [comment]
I'm putting this on the tacit resolution queue for next meeting (although I'll be OOO). Here's a summary for the meeting:

---

The current spec says imported video textures are closed very quickly (in a microtask after the current task). This leads to developers needing to reimport frequently, which makes them feel they should cache the data instead of reimporting (even though copying is more expensive than importing).

This PR extends the lifetime until the frame is not current on the source video anymore. For predictability, expiry always happens at a very specific point - within a frame after the video advances, the lifetime ends immediately before requestVideoFrameCallback callbacks, which are just before requestAnimationFrame callbacks.

It also adds a boolean attribute `GPUExternalTexture.expired` which tells the app to reimport to get a new frame. Apps use this if WebGPU rendering isn't driven by requestVideoFrameCallback already. (Could be approximated by observing requestVideoFrameCallback, but that's a bit indirect, slightly imprecise, and not in all browsers. This is direct and easy to use.)

**Review by kdashg — APPROVED** [review]

**Comment by kdashg** [comment]
details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-2022-02-16#tacit-resolution-queue">WebGPU meeting minutes 2022-02-16</a></summary>

* [ Extend lifetime of GPUExternalTexture imported from video element #2302 ](https://github.com/gpuweb/gpuweb/pull/2302) 
    * KN: Importing external textures from video elements
    * Conservatively spec'ed initially. Textures would expire quickly.
    * Devs need to re-import the same frame multiple times. Looks expensive, even though we'd cache inside the browser.
    * Extends the lifetime in a well-defined way - until frame's no longer current.
    * Expires texture at very specific time in frame's lifetime.
    * Think this is a fairly safe change, and better constraint than what we had previously. Should work better for programming against.
    * Does add a boolean attribute to GPUExternalTexture saying when it's expired. Transitions at a well-defined point.
    * KG: this is at some microtask boundary?
    * KN: yes. Right before requestAnimationFrame, or if you have requestVideoFrameCallback - it's right before that.
    * MM: if I import a video frame, and next line does the same thing - guaranteed to get the same frame?
    * KN: I don't think so. Not 100% sure. Videos behave asynchronously, change behind the scenes.
    * MM: what happens in WebGL if you texImage2D twice in a row?
    * KG / KN: not guaranteed to get the same frame. Same guarantees here.
    * KG: if you had 1000 fps video for example you could theoretically get two different frames.
    * BJ: "expires" flag and requestVideoFrame loop aren't necessarily ticking for each frame in this case.
    * KN: if you imported twice you could get 2 different objects that expire at 2 different times.
    * MM: memory use would be proportional to frame rate of video if you imported a bunch of times.
    * KG: potentially.
    * BJ: could you observe in a lightweight way that you did get a different object back? Guaranteed to get same JS object back if frame didn't change?
    * KN: pretty sure you get a new JS object.
    * BJ: imagining blog post where someone tries to determine frame rate of video by repeatedly re-fetching frames.
    * Discussion about frame rates, time stamps of frames.
    * KG: texture uploads from video are naive in WebGL - tried to provide additional information like width, height, timestamp - didn't end up finishing those proposals.
    * KN: and WebCodecs came along while implementing this, and VideoFrame handles this case.
    * KN: concerns with landing this?
    * MM: no - just think some more design work may be needed to ensure the tutorial Brandon hypothesized doesn't get written.
    * MM: one more question - there's requestVideoFrame - will these handles expire at the next 60 Hz rAF or the next frame of the video they came from?
    * KN: the latter. Doesn't expire until after the video frame isn't current any more. At next boundary when video advances to next frame - the GPUExternalTexture will expire. Last and current external textures' lifetimes may overlap slightly.
    * KG: think we're good to land this.
    * MM: this is good, you can write a naive requestVideoFrame loop and it'll work.
    * KN: yes. Can drive rendering both by video frame rate and animation frame rate. Works well in both cases.


</details>

**Comment by lokokung** [comment]
Removing 'needs-cts-issue' for now because this is obsolete and will be updated in a future pull request to fix #3324.
